File: src/main/scala/conway/Board.scala
package conway


case class Board(rows: Int, columns: Int, creatures: Set[Creature]) {

	// secondary "index" of creatures for optimized containment lookup
	private val grid: Array[Array[Creature]] = new Array(rows)

	// Instance initialization
	creatures.find(c => c.row > rows - 1 || c.column > columns - 1) match {
		case Some(c) =>
			throw new IllegalArgumentException(s"Creature off-board: $c [board: ${this.rows} x ${this.columns}]")
		case _ =>
	}

	// Build the grid of cells, and populate with creatures
	for (rowNumber <- 0 until rows) {
		grid(rowNumber) = new Array[Creature](columns)
	}
	creatures.foreach(creature => grid(creature.row)(creature.column) = creature)


	def containsCreature(coordinate: Coordinate): Boolean = {
		// Early out if the query is off-board
		if (coordinate.row < 0 || coordinate.row >= rows) {
			return false
		}
		if (coordinate.column < 0 || coordinate.column >= columns) {
			return false
		}

		grid(coordinate.row)(coordinate.column) != null
	}
}

File: java8/src/main/java/conway/java/Board.java
package conway.java;

import java.util.Set;
import java.util.stream.IntStream;

public class Board {

	private final int rows;
	private final int columns;
	private final Set<Creature> creatures;
	// secondary "index" of creatures for optimized containment lookup
	private final Creature[][] grid;

	public Board(int rows, int columns, Set<Creature> creatures) {
		this.rows = rows;
		this.columns = columns;
		this.creatures = creatures;

		grid = new Creature[rows][];
		IntStream.range(0, rows).forEach(r -> grid[r] = new Creature[columns]);

		creatures.stream()
				.forEach(c -> {
					if (c.getRow() > rows - 1 || c.getColumn() > columns - 1) {
						String msg = String.format("Creature off-board: %s [board: %d x %d]", c, rows, columns);
						throw new IllegalArgumentException(msg);
					}
					grid[c.getRow()][c.getColumn()] = c;
				});
	}

	public boolean creatureAtCoordinate(Coordinate coordinate) {
		return containsCreature(coordinate.getRow(), coordinate.getColumn());
	}

	public boolean creatureRelativeToCoordinate(Coordinate coordinate, Offset offset) {
		return containsCreature(
				coordinate.getRow() + offset.getRowDelta(),
				coordinate.getColumn() + offset.getColumnDelta()
		);
	}

	public int getRows() {
		return rows;
	}

	public int getColumns() {
		return columns;
	}

	public Set<Creature> getCreatures() {
		return creatures;
	}

	private boolean containsCreature(int row, int column) {
		return row >= 0 && row < rows &&
				column >= 0 && column < columns &&
				grid[row][column] != null;
	}
}
